stdlib: excise the FP16 support routines on x64

compnerd opened the pull request:
Bump the minimum supported CPU to ~Nahelem so that we can take advantage of the F16C/CVT16 extensions to support half floating point rounding. This should avoid the need to replicate the compiler-rt functionality in the runtime and the associated problems with linking compiler-rt builtins on non-Windows targets when performing static linking.

compnerd committed:
X86: bump the minimum required CPU to IvyBridge

Bump the minimum supported CPU to IvyBridge so that we can take
advantage of the F16C/CVT16 extensions to support half floating point
rounding.  This avoids the undefined reference to `__extendsfhf2` in the
FP16 support.

compnerd commented:
CC: @stephentyrone @airspeedswift @bnbarham @etcwilde @eeckstein

bnbarham commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -778,6 +778,12 @@ importer::addCommonInvocationArguments(
     // Enable double wide atomic intrinsics on every x86_64 target.
     // (This is the default on Darwin, but not so on other platforms.)
     invocationArgStrs.push_back("-mcx16");
+    // Enable the FC16/CVT16 extensions which provide half prevision floating
```
```suggestion
    // Enable the FC16/CVT16 extensions which provide half precision floating
```

bnbarham commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -288,6 +288,18 @@ function(_add_target_variant_c_compile_flags)
+  # Avoid the need for the FP16 truncation and extension routines from
+  # compiler-rt by assuming that we have hardware capable of performing
+  # half-precision floating point conversions.  This was announced circa 2007,
+  # and is standard with ~Nahelem (~2013).
```
```suggestion
  # and is standard with ~Haswell (~2013).
```

IIUC

compnerd replied:
Yes, sorry about that!  Fixed that locally.

compnerd commented:
CC: @rjmccall

buttaface commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -288,6 +288,18 @@ function(_add_target_variant_c_compile_flags)
+    else()
+      list(APPEND result -mf16c)
+    endif()
+  endif()
```
Why not simply combine this with the x86_64 block just above?

compnerd replied:
We absolutely could, it was more a matter of convenience and isolation.  Having it be interleaved is slightly annoying to update in the future.  I'm hoping that as we try to clean up the CMake rules, we can get this all cleaned up.

buttaface commented on stdlib/public/runtime/Float16Support.cpp:
```diff
@@ -29,8 +29,7 @@
 // provide shims in that case.
-#if (defined(__ANDROID__) && defined(__ARM_ARCH_7A__) && defined(__ARM_EABI__)) || \
-  ((defined(__i386__) || defined(__i686__) || defined(__x86_64__)) && !defined(__APPLE__) && !defined(__linux__))
+#if (defined(__ANDROID__) && defined(__ARM_ARCH_7A__) && defined(__ARM_EABI__))
```
Would this break this on Win32 and some other 32-bit x86 platforms? I don't build or use i686 anymore, so don't really care myself, but may be something to watch out for.

compnerd replied:
I don't think that this will break the Windows x86 builds if we were to get the x86_64 builds complete.  The CPU is still the same, it is just the bitness that changes.  It will however be a bit messier codegen due to the calling convention.

stephentyrone commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -778,6 +778,12 @@ importer::addCommonInvocationArguments(
     invocationArgStrs.push_back("-mcx16");
+    // Enable the FC16/CVT16 extensions which provide half precision floating
+    // point support on x86_64.  This bumps the minimum requirement of the CPU
+    // to ~Haswell, however, Windows 10 (at least as of 1709) requires
```
F16C was introduced with Ivybridge, not Haswell.

compnerd replied:
Ivybridge just preceded Haswell on the tick IIRC, but I can absolutely update the comments to be more precise (and I do like precision).  Thanks for the history check @stephentyrone!

stephentyrone commented:
This doesn’t actually solve the problem on its own; F16C provides conversions float <-> float16, but the compiler will still generate runtime calls for conversions to/from double and float80. I’ve worked around these in the stdlib for the conversions that were missing in macOS’s compiler-rt, but there are some others IIRC.

compnerd commented:
@stephentyrone - that is good to know!  Do you have an example of that?  I am hitting another issue in the compiler atm, so I haven't gotten to that point yet.  At least via testing through clang, it appeared that the truncation and extension would get lowered to F16C.

compnerd commented:
@swift-ci please test Windows platform

shahmishal commented:
@swift-ci test

stephentyrone commented:
@compnerd here's an example of a runtime call generated even with f16c (by clang): https://godbolt.org/z/6b7513on6

stephentyrone commented:
https://godbolt.org/z/qsYqTvb1c

stephentyrone commented:
I can work around that one in the stdlib for you, but more generally we need to be able to use the arithmetic builtins from the just-built compiler-rt, so making that work is the long-term solution we really want.

compnerd commented:
Most of the math routines shouldn't be forming libcalls on Windows.  FP16 operations are odd since there is no FP16 on Windows AFAIK.  We should be forcing LLVM to lower that properly in the longer term rather than having to workaround it at the library level.

stephentyrone commented:
I'm not sure what "forcing LLVM to lower that properly" means other than generating a libcall; that's the proper lowering.

compnerd commented:
I was thinking that it should avoid the libcall on Windows specifically - as the OS doesn't expect to link compiler-rt ever (it should be similar to what MSVC does).

However, a trophy for you - you correctly identified that `__truncdfhf2` is needed after the X86 CodeGen issue is resolved.

compnerd commented:
Please test with following PRs:
apple/llvm-project#5995

@swift-ci please test

compnerd commented:
Please test with following PRs:
https://github.com/apple/llvm-project/pull/5995

@swift-ci please test Linux platform

compnerd commented:
@swift-ci please test Windows platform

stephentyrone commented on test/IRGen/ordering_x86.sil:
```diff
@@ -41,4 +41,4 @@ bb0:
 
 // X86_64: define{{( protected)?}} swiftcc void @baz{{.*}}#0
-// X86_64: #0 = {{.*}}"target-features"="+cx16,+cx8,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+ssse3,+x87"
+// X86_64: #0 = {{.*}}"target-features"="+avx,+crc32,+cx16,+cx8,+f16c,+fxsr,+mmx,+popcnt,+sahf,+sse,+sse2,+sse3,+sse4.1,+sse4.2,+ssse3,+x87,+xsave"
```
We're definitely not OK with bumping the baseline requirement for swift on x86 by ~4 years without a language workgroup and/or core team discussion. I think we should really try to fix this in a different manner.

compnerd replied:
Yeah, I think that the Language WG should be involved in that - I think that @airspeedswift was relayed this though, and @rjmccall was also CC'ed for that.

compnerd replied:
I should also mention that the f16c removes the `__extendhfsf2` not the `__truncsfhf2` and that the latter is something that I will restore for the time being.  I think that one option might be to add `__extendhfsf2` to the support routines.

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
Please test with following PRs:
apple/swift-corelibs-foundation#4689

@swift-ci please build toolchain Windows platform

compnerd commented:
Please test with following PRs:
apple/swift-corelibs-foundation#4689

@swift-ci please build toolchain Windows platform

compnerd commented:
@swift-ci please build toolchain Windows platform

compnerd commented:
@swift-ci please build toolchain Windows platform

compnerd commented:
```
-- Testing: 8865 tests, 36 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 

3 warning(s) in tests

Testing Time: 390.24s
  Unsupported      : 2104
  Passed           : 6884
  Expectedly Failed:   51
```

```
Test project T:/3
      Start  1: dispatch_apply
 1/22 Test  #1: dispatch_apply ...................   Passed    0.18 sec
      Start  2: dispatch_api
 2/22 Test  #2: dispatch_api .....................   Passed    0.14 sec
      Start  3: dispatch_debug
 3/22 Test  #3: dispatch_debug ...................   Passed    0.14 sec
      Start  4: dispatch_queue_finalizer
 4/22 Test  #4: dispatch_queue_finalizer .........   Passed    1.14 sec
      Start  5: dispatch_overcommit
 5/22 Test  #5: dispatch_overcommit ..............   Passed    0.15 sec
      Start  6: dispatch_context_for_key
 6/22 Test  #6: dispatch_context_for_key .........   Passed    0.14 sec
      Start  7: dispatch_after
 7/22 Test  #7: dispatch_after ...................   Passed    9.16 sec
      Start  8: dispatch_timer
 8/22 Test  #8: dispatch_timer ...................   Passed    2.15 sec
      Start  9: dispatch_timer_short
 9/22 Test  #9: dispatch_timer_short .............   Passed    1.40 sec
      Start 10: dispatch_timer_timeout
10/22 Test #10: dispatch_timer_timeout ...........   Passed    6.15 sec
      Start 11: dispatch_sema
11/22 Test #11: dispatch_sema ....................   Passed    0.22 sec
      Start 12: dispatch_timer_bit31
12/22 Test #12: dispatch_timer_bit31 .............   Passed    2.30 sec
      Start 13: dispatch_timer_bit63
13/22 Test #13: dispatch_timer_bit63 .............   Passed    1.15 sec
      Start 14: dispatch_timer_set_time
14/22 Test #14: dispatch_timer_set_time ..........   Passed    2.45 sec
      Start 15: dispatch_data
15/22 Test #15: dispatch_data ....................   Passed    0.15 sec
      Start 16: dispatch_io_muxed
16/22 Test #16: dispatch_io_muxed ................   Passed    1.36 sec
      Start 17: dispatch_io_net
17/22 Test #17: dispatch_io_net ..................   Passed    0.47 sec
      Start 18: dispatch_io_pipe
18/22 Test #18: dispatch_io_pipe .................   Passed   10.17 sec
      Start 19: dispatch_io_pipe_close
19/22 Test #19: dispatch_io_pipe_close ...........   Passed    0.16 sec
      Start 20: dispatch_select
20/22 Test #20: dispatch_select ..................   Passed    0.22 sec
      Start 21: dispatch_c99
21/22 Test #21: dispatch_c99 .....................   Passed    0.15 sec
      Start 22: dispatch_plusplus
22/22 Test #22: dispatch_plusplus ................   Passed    0.15 sec

100% tests passed, 0 tests failed out of 22

Total Test time (real) =  39.81 sec
```

```
The following tests FAILED:
	1281 - TestFoundation.TestNSString-test_NSHomeDirectoryForUser (Failed)
	1283 - TestFoundation.TestNSString-test_expandingTildeInPath (Failed)
	1347 - TestFoundation.TestProcess-test_multiProcesses (Failed)
	1377 - TestFoundation.TestURL-test_URLByResolvingSymlinksInPathShouldUseTheCurrentDirectory (Failed)
	1378 - TestFoundation.TestURL-test_resolvingSymlinksInPathShouldAppendTrailingSlashWhenExistingDirectory (Failed)
	1379 - TestFoundation.TestURL-test_resolvingSymlinksInPathShouldResolveSymlinks (Failed)
```

```
[0/1][  0%][0.000s] Running XCTest functional test suite
-- Testing: 26 tests, 26 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 

Testing Time: 24.11s
  Passed: 26
```

The Foundation test failures are intriguing as locally, I am seeing only 2 failures.  But seems that Swift, libdispatch, and XCTest are good.  The Foundation test failures seem to match the current ones in CI, so I think that Windows is still looking good assuming we figure out the FP16 issue.

compnerd commented:
```
swift-corelibs-foundation\Tests\Foundation\Tests\TestNSString.swift:1009: error: TestNSString.test_NSHomeDirectoryForUser :
  XCTAssertEqual failed: ("Optional("C:/Users/swift-ci")") is not equal to ("nil") - 

swift-corelibs-foundation\Tests\Foundation\Tests\TestNSString.swift:1036: error: TestNSString.test_expandingTildeInPath :
  XCTAssertEqual failed: ("~swift-ci") is not equal to ("C:/Users/swift-ci") - Could resolve home directory for specific user

swift-corelibs-foundation\Tests\Foundation\Tests\TestProcess.swift:803: error: TestProcess.test_multiProcesses :
  XCTAssertEqual failed: ("T:/4/TestFoundation.app") is not equal to ("C:/Users/swift-ci/jenkins/workspace/swift-PR-build-toolchain-windows/build/4/TestFoundation.app") - 

swift-corelibs-foundation\Tests\Foundation\Tests\TestURL.swift:562: error: TestURL.test_URLByResolvingSymlinksInPathShouldUseTheCurrentDirectory :
  XCTAssertEqual failed: ("file:///C:/Users/swift-ci/jenkins/workspace/swift-PR-build-toolchain-windows/build/tmp/org.swift.TestFoundation.TestURL.resourceValues.464/foo/bar/baz") is not equal to ("file:///T:/tmp/org.swift.TestFoundation.TestURL.resourceValues.464/foo/bar/baz") - 

swift-corelibs-foundation\Tests\Foundation\Tests\TestURL.swift:576: error: TestURL.test_resolvingSymlinksInPathShouldAppendTrailingSlashWhenExistingDirectory :
  XCTAssertEqual failed: ("file:///C:/Users/swift-ci/jenkins/workspace/swift-PR-build-toolchain-windows/build/tmp/org.swift.TestFoundation.TestURL.resourceValues.7768/") is not equal to ("file:///T:/tmp/org.swift.TestFoundation.TestURL.resourceValues.7768/") - 

swift-corelibs-foundation\Tests\Foundation\Tests\TestURL.swift:591: error: TestURL.test_resolvingSymlinksInPathShouldResolveSymlinks :
  XCTAssertEqual failed: ("file:///C:/Users/swift-ci/jenkins/workspace/swift-PR-build-toolchain-windows/build/tmp/org.swift.TestFoundation.TestURL.resourceValues.5012/destination") is not equal to ("file:///T:/tmp/org.swift.TestFoundation.TestURL.resourceValues.5012/destination") - 
```

Some text editing with the actual failures here indicates that the failures are the two that I see locally, and 4 related to symlink handling where the expectation doesn't match due to the build being on a different drive than the source.

compnerd commented:
@swift-ci please build toolchain Windows platform

bnbarham commented:
@swift-ci please test Linux platform

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd committed:
runtime: add `__extendhfxf2` implementation

Add a simple special case implementation of the FP16 routines for the FP80 extension.

Implementation by @al45tair!

compnerd commented:
@swift-ci please test

compnerd committed:
Update Float16Support.cpp

compnerd commented:
@swift-ci please test

compnerd committed:
Update Float16Support.cpp

compnerd commented:
@swift-ci please test

compnerd committed:
Update Float16Support.cpp

compnerd commented:
@swift-ci please test

compnerd committed:
Update Float16Support.cpp

compnerd commented:
@swift-ci please test Linux platform

compnerd committed:
Update Float16Support.cpp

compnerd commented:
@swift-ci please test Linux platform

bnbarham commented:
@swift-ci please test Linux platform

compnerd committed:
Update Float16Support.cpp

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
I think that this is no longer needed!

lin72h commented:
> 

I've been watching the progress on it, can you elaborate more on the reason?

etcwilde commented:
> I've been watching the progress on it, can you elaborate more on the reason?

@al45tair fixed it without requiring bumping the minimum CPU version in https://github.com/apple/swift/commit/125e54f38cef27bd539b18221b94b32ff73e6e1f and https://github.com/apple/swift/commit/df1891eb983690d18f029d0fcf38b8b423729199. Not having to bump the minimum processor version is desirable.

lin72h commented:
@etcwilde perfect, Thanks
